watson_developer_cloud/iam_token_manager.py

```diff
@@ -0,0 +1,142 @@
+import requests
+import time
+
+DEFAULT_IAM_URL = 'https://iam.bluemix.net/identity/token'
+CONTENT_TYPE = 'application/x-www-form-urlencoded'
+ACCEPT = 'application/json'
+DEFAULT_AUTHORIZATION = 'Basic Yng6Yng='
+REQUEST_TOKEN_GRANT_TYPE = 'urn:ibm:params:oauth:grant-type:apikey'
+REQUEST_TOKEN_RESPONSE_TYPE = 'cloud_iam'
+REFRESH_TOKEN_GRANT_TYPE = 'refresh_token'
+
+class IAMTokenManager(object):
+    def __init__(self, iam_api_key=None, iam_access_token=None, iam_url=None):
+        self.iam_api_key = iam_api_key
+        self.user_access_token = iam_access_token
+        self.url = iam_url if iam_url else DEFAULT_IAM_URL
+        self.token_info = {
+            'access_token': None,
+            'refresh_token': None,
+            'token_type': None,
+            'expires_in': None,
+            'expiration': None,
+        }
+
+    def request(self, method, url, headers=None, params=None, data=None, **kwargs):
+        response = requests.request(method=method, url=url,
+                                    headers=headers, params=params,
+                                    data=data, **kwargs)
+        if 200 <= response.status_code <= 299:
+            return response.json()
+        else:
+            from .watson_service import WatsonApiException, get_error_message
+            error_message = get_error_message(response)
+            raise WatsonApiException(response.status_code, message=error_message, httpResponse=response)
+
+    def get_token(self):
+        """
+        The source of the token is determined by the following logic:
+        1. If user provides their own managed access token, assume it is valid and send it
+        2. If this class is managing tokens and does not yet have one, make a request for one
+        3. If this class is managing tokens and the token has expired refresh it. In case the refresh token is expired, get a new one
+        If this class is managing tokens and has a valid token stored, send it
+        """
+        if self.user_access_token:
+            return self.user_access_token
+        elif not self.token_info.get('access_token'):
+            token_info = self.request_token()
+            self.save_token_info(token_info)
+            return self.token_info.get('access_token')
+        elif self.is_token_expired():
+            if self.is_refresh_token_expired():
+                token_info = self.request_token()
+            else:
+                token_info = self.refresh_token()
+            self.save_token_info(token_info)
+            return self.token_info.get('access_token')
+        else:
+            return self.token_info.get('access_token')
+
+    def request_token(self):
```

Comment:
This is a private method

Reply:
Changing the following to private by adding a prefix `_`:
_request_token
_refresh_token
_is_token_expired
_is_refresh_token_expired
_save_token_info